500 ISE for Client Side Error

akka-http M5

If you submit a HEAD request which has an entity body, you get back a 500 Internal Server Error.  The exception that triggers this is a require stating "Requests with this method must have an empty entity".

This requirement can only be satisfied by a client however, so a 4xx series error seems more appropriate here?

```
cc2 [ERROR] [04/13/2015 16:01:07.698] [cloud-controller-akka.actor.default-dispatcher-27] [ActorSystem(cloud-controller)] Internal server error, sending 500 response
cc2 java.lang.IllegalArgumentException: requirement failed: Requests with this method must have an empty entity
cc2 	at scala.Predef$.require(Predef.scala:219)
cc2 	at akka.http.model.HttpRequest.<init>(HttpMessage.scala:133)
cc2 	at akka.http.engine.server.HttpServer$$anonfun$6.apply(HttpServer.scala:87)
cc2 	at akka.http.engine.server.HttpServer$$anonfun$6.apply(HttpServer.scala:83)
cc2 	at akka.stream.impl.fusing.Map.onPush(Ops.scala:16)
cc2 	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$1.run(Interpreter.scala:317)
cc2 	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.progress(Interpreter.scala:224)
cc2 	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$1.progress(Interpreter.scala:315)
cc2 	at akka.stream.impl.fusing.OneBoundedInterpreter.akka$stream$impl$fusing$OneBoundedInterpreter$$execute(Interpreter.scala:447)
cc2 	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$6.push(Interpreter.scala:535)
cc2 	at akka.stream.impl.fusing.BatchingActorInputBoundary$$anonfun$upstreamRunning$1.applyOrElse(ActorInterpreter.scala:139)
cc2 	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
cc2 	at akka.stream.impl.SubReceive.apply(Transfer.scala:16)
cc2 	at akka.stream.impl.SubReceive.apply(Transfer.scala:12)
cc2 	at scala.PartialFunction$class.applyOrElse(PartialFunction.scala:123)
cc2 	at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:12)
cc2 	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:170)
cc2 	at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
cc2 	at akka.stream.impl.fusing.ActorInterpreter.aroundReceive(ActorInterpreter.scala:310)
cc2 	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
cc2 	at akka.actor.ActorCell.invoke_aroundBody0(ActorCell.scala:487)
cc2 	at akka.actor.ActorCell$AjcClosure1.run(ActorCell.scala:1)
cc2 	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
cc2 	at akka.kamon.instrumentation.ActorCellInstrumentation$$anonfun$aroundBehaviourInvoke$1.apply(ActorCellInstrumentation.scala:62)
cc2 	at kamon.trace.Tracer$.withContext(Tracer.scala:56)
cc2 	at akka.kamon.instrumentation.ActorCellInstrumentation.aroundBehaviourInvoke(ActorCellInstrumentation.scala:61)
cc2 	at akka.actor.ActorCell.invoke(ActorCell.scala:483)
cc2 	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
cc2 	at akka.dispatch.Mailbox.run(Mailbox.scala:221)
cc2 	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
cc2 	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
cc2 	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
cc2 	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
cc2 	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpRequestParser.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpRequestParser.scala
@@ -167,6 +167,8 @@ private[http] class HttpRequestParser(_settings: ParserSettings,
             emitRequestStart(emptyEntity(cth))
             setCompletionHandling(HttpMessageParser.CompletionOk)
             startNewMessage(input, bodyStart)
+          } else if (!method.isEntityAccepted) {
+            failMessageStart(UnprocessableEntity, s"${method.name} requests must not have an entity")
           } else if (contentLength <= input.size - bodyStart) {
             val cl = contentLength.toInt
             emitRequestStart(strictEntity(cth, input, bodyStart, cl))
@@ -176,6 +178,9 @@ private[http] class HttpRequestParser(_settings: ParserSettings,
             emitRequestStart(defaultEntity(cth, contentLength, expect100continueHandling))
             parseFixedLengthBody(contentLength, closeAfterResponseCompletion)(input, bodyStart)
           }
+
+        case Some(_) if !method.isEntityAccepted ⇒
+          failMessageStart(UnprocessableEntity, s"${method.name} requests must not have an entity")
 
         case Some(te) ⇒
           val completedHeaders = addTransferEncodingWithChunkedPeeled(headers, te)

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/parsing/RequestParserSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/parsing/RequestParserSpec.scala
@@ -138,11 +138,11 @@ class RequestParserSpec extends FreeSpec with Matchers with BeforeAndAfterAll {
       }
 
       "with a funky `Transfer-Encoding` header" in new Test {
-        """GET / HTTP/1.1
+        """PUT / HTTP/1.1
           |Transfer-Encoding: foo, chunked, bar
           |Host: x
           |
-          |""" should parseTo(HttpRequest(GET, "/", List(`Transfer-Encoding`(TransferEncodings.Extension("foo"),
+          |""" should parseTo(HttpRequest(PUT, "/", List(`Transfer-Encoding`(TransferEncodings.Extension("foo"),
           TransferEncodings.chunked, TransferEncodings.Extension("bar")), Host("x"))))
         closeAfterResponseCompletion shouldEqual Seq(false)
       }
@@ -398,6 +398,19 @@ class RequestParserSpec extends FreeSpec with Matchers with BeforeAndAfterAll {
             |Host: x
             |
             |""" should parseToError(400: StatusCode, ErrorInfo("`Content-Length` header value must not exceed 63-bit integer range"))
+        }
+
+        "with an illegal entity" in new Test {
+          """HEAD /resource/yes HTTP/1.1
+            |Content-length: 3
+            |Host: x
+            |
+            |foo""" should parseToError(422: StatusCode, ErrorInfo("HEAD requests must not have an entity"))
+          """DELETE /resource/yes HTTP/1.1
+            |Transfer-Encoding: chunked
+            |Host: x
+            |
+            |""" should parseToError(422: StatusCode, ErrorInfo("DELETE requests must not have an entity"))
         }
       }
     }

